lib/calendars.ml

```diff
@@ -251,194 +303,226 @@ let sdn_of_hebrew d =
                 in
                 let tishri1 = fTishri1 metonicYear day halakim in
-                let (month, day) = glop inputDay tishri1 tishri1After in
-                year, month, day
+                let month, day = glop inputDay tishri1 tishri1After in
+                (year, month, day)
   in
-  {day = day; month = month; year = year; delta = 0}
-
-(* from and to gregorian *)
-
-let conv f f_max_month g g_max_month d =
-  let sdn =
-    if d.day = 0 then
-      if d.month = 0 then g {d with day = 1; month = 1}
-      else g {d with day = 1}
-    else g d
+  { day; month; year; delta = 0; kind = Hebrew }
+
+let kind_to_string : type a. a kind -> string =
+ fun kind ->
+  match kind with
+  | Gregorian -> "Gregorian"
+  | Julian -> "Julian"
+  | French -> "French"
+  | Hebrew -> "Hebrew"
+
+let make :
+    type a.
+    a kind ->
+    day:int ->
+    month:int ->
+    year:int ->
+    delta:int ->
+    (a date, string) result =
+ fun kind ~day ~month ~year ~delta ->
+  let gregorian_nb_days_upper_bound =
+    [| 31; 29; 31; 30; 31; 30; 31; 31; 30; 31; 30; 31 |]
   in
-  let sdn_max =
-    if d.day = 0 then
-      if d.month = 0 || d.month = g_max_month then
-        g {day = 1; month = 1; year = d.year + 1; delta = 0}
-      else
-        g
-          {day = 1; month = d.month + 1; year = d.year;
-           delta = 0}
-    else sdn + 1
+  let hebrew_nb_days_upper_bound =
+    (* last two are for Adar I and II;
+       Adar II should be 29, but we keep it simple and do not check for leap years *)
+    [| 30; 29; 30; 29; 30; 29; 30; 30; 30; 29; 30; 30; 30 |]
   in
-  let d1 = f sdn in
-  let d2 = f (sdn_max + d.delta) in
-  if d1.day = 1 && d2.day = 1 then
-    if d1.month = 1 && d2.month = 1 then
-      if d1.year + 1 = d2.year then
-        {day = 0; month = 0; year = d1.year; delta = 0}
-      else {d1 with delta = sdn_max + d.delta - sdn - 1}
-    else if
-      d1.month + 1 = d2.month ||
-      d1.month = f_max_month && d1.year + 1 = d2.year
-    then
-      {d1 with day = 0}
-    else {d1 with delta = sdn_max + d.delta - sdn - 1}
-  else {d1 with delta = sdn_max + d.delta - sdn - 1}
-
-let gregorian_of_julian = conv gregorian_of_sdn 12 sdn_of_julian 12
-let julian_of_gregorian = conv julian_of_sdn 12 sdn_of_gregorian 12
-
-let gregorian_of_french = conv gregorian_of_sdn 12 sdn_of_french 13
-let french_of_gregorian = conv french_of_sdn 13 sdn_of_gregorian 12
-
-let gregorian_of_hebrew = conv gregorian_of_sdn 12 sdn_of_hebrew 13
-let hebrew_of_gregorian = conv hebrew_of_sdn 13 sdn_of_gregorian 12
+  let is_invalid =
+    day < 1 || month < 1
+    ||
+    match kind with
+    | Gregorian | Julian ->
+        (* from wikipedia: "A year zero does not exist in the Anno Domini (AD) calendar year
+           system commonly used to number years in the Gregorian calendar (nor in its
+           predecessor, the Julian calendar)" *)
+        year = 0
+        || (match kind with
+           | Gregorian -> true
```

Comment:
I don't understand what's going on here. A Gregorian date is always going to be invalid ?